### PR TITLE
DrivAerML: 5L/512d + gc=1.0 (stabilized deeper architecture)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

6L/512d with gc=0.5 reached 6.372% (PR #3078) but diverged at ep155. The monotonic depth finding (2L<3L<4L) was reversed at 6L. The untested point is 5L — if 5L sits between 4L (3.997%) and 6L (6.372%), it could beat the baseline.

PR #3077 (gilbert) is testing 5L WITHOUT gc. This PR tests 5L WITH gc=1.0 to stabilize it for 400+ epoch training.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 5 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm-5l-gc
```

Report best val_primary/surface_rel_l2_pct and divergence epoch if applicable.

## Baseline
- **val_primary/surface_rel_l2_pct:** 3.997% at epoch 467 (4L, no gc)
- **6L + gc=0.5 result:** 6.372% (diverged ep155) — depth above 4L hurts at standard config

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT vegeta:

## Results

### 5L/512d + gc=1.0 DrivAerML

| Metric | Baseline (4L/512d) | This PR (5L/512d gc=1.0) | Delta |
|---|---|---|---|
| **val_primary/surface_rel_l2_pct** | **3.997%** (ep467) | **5.515%** (ep167) | **+38.0% (WORSE)** |
| Divergence epoch | N/A (stable 467+ eps) | **ep180** | — |

**W&B run:** w7e9yqgw (vegeta/dm-5l-gc10)

**Command:**
\`\`\`bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 5 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent vegeta --wandb-name "vegeta/dm-5l-gc10" --wandb-group dm-5l-gc
\`\`\`

**Peak memory:** ~7.1 GB VRAM (GPU 0)

### What happened

The 5L/512d model with gc=1.0 **does NOT beat the 4L/512d baseline**. The hypothesis that 5L would sit between 4L (3.997%) and 6L (6.372%) is **partially confirmed** in ordering — 5L (5.515%) is indeed between 4L and 6L — but it still significantly underperforms the 4L baseline.

**Training trajectory:**
- Epochs 1-60: Rapid descent from 60% to ~10%, consistent trough-to-trough improvement
- Epochs 60-167: Continued descent, best 5.515% at ep167 (cosine trough)
- **Epoch 180: Catastrophic divergence** — grad norms exploded from 0.5 to 23.6 in a single epoch, then to infinity by ep250
- Post-divergence: Model collapsed to ~73.4% (random-level), grad norms at infinity

**Divergence analysis:** gc=1.0 was insufficient to prevent the instability. At ep179 (one epoch before divergence), grad_norm_mean jumped to 0.50 from the typical ~0.28 range, then at ep180 loss spiked from 0.013 to 0.226 with grad_norm=23.6. This matches the same T_max=30 cosine peak instability pattern seen in the 6L run (diverged ep155). The extra layer amplifies gradient variance at cosine LR peaks.

**Depth-accuracy ordering on DrivAerML (all 512d, T_max=30):**
- 4L: 3.997% at ep467 (stable, no divergence)
- 5L + gc=1.0: 5.515% at ep167 (diverged ep180)
- 6L + gc=0.5: 6.372% (diverged ep155)

**Conclusion:** Depth beyond 4L hurts on DrivAerML. The monotonic depth benefit seen at 2L→3L→4L reverses at 5L+. The extra depth increases gradient variance, making training less stable even with gradient clipping, and the model doesn't find as deep a basin before divergence.

### Suggested follow-ups

1. **5L + gc=0.5 or gc=0.3:** Tighter clipping might delay divergence long enough for 5L to match or beat 4L, though the earlier 6L+gc=0.5 result (still diverged at ep155) suggests this is unlikely
2. **5L + EMA=0.999:** EMA could preserve the best checkpoint through instability and potentially find deeper basins, analogous to TandemFoil results
3. **5L + lower LR (e.g. 3e-4):** Reducing LR might stabilize the deeper architecture enough to train longer
4. **Abandon depth >4L on DrivAerML:** The evidence is building that 4L is the optimal depth for this dataset — focus efforts on other hyperparameters instead

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3141 because ADVISOR: 5L+gc=1.0 reaches 5.515% (ep167) then diverges ep180. Three independent 5L tests now confirm depth>4L is dead: 5L no-gc (#3077, 4.172%), 5L+gc=1.0 (this, 5.515%), 6L (#3078, 6.37%). Extra layers amplify gradient variance at cosine LR peaks, triggering earlier divergence regardless of gc. 4L/512d is the optimal DrivAerML architecture. Closing..
